### DISCUSSION TITLE
yabai comparable eventing - additional app/window/display events with corresponding IDs

### DISCUSSION THREAD

**Discussion by kenbanks-peng** [nikitabobko/AeroSpace#783] [feature-ideas]
I am using sketchybar to not only track spaces, but to track windows with icons. I can highlight the active space using the aerospace_workspace_changed event, but I can highlight the focused app with yabai events. 

presently using yabai in float layout (uncontrolled, to avoid conflict with aerospace), just so I can get access to the more granular eventing. 

would request aerospace eventing be granualar as well

some yabai events:
- application_launched
- application_terminated
- application_front_switched
- application_activated
- application_deactivated
- application_visible
- application_hidden
- window_created
- window_destroyed
- window_focused
- window_moved
- window_resized
- window_minimized
- window_deminimized
- window_title_changed
- display_added
- display_removed
- display_moved
- display_resized
- display_changed

**Comment by mehd-io** [comment]
Second this, it's the exact only reason why I'm having hard time to migrate completely out of yabai.
Having visibility on running apps/windows on each spaces is must, but I couldn't figure out any callback way to trigger an update on sketchybar when moving an app to another workspace or quitting one.

**Comment by kenbanks-peng** [comment]
I certainly could be overlooking something but here are the yabai services that I am still relying on. Aerospace seems focused on simulated spaces, where yabai brings in window/app events.

yabairc
```
yabai -m signal --add event=window_created action="sketchybar --trigger yabai_window_created ID=\$YABAI_WINDOW_ID &> /dev/null"
yabai -m signal --add event=window_destroyed action="sketchybar --trigger yabai_window_destroyed ID=\$YABAI_WINDOW_ID &> /dev/null"
yabai -m signal --add event=window_moved action="sketchybar --trigger yabai_window_moved ID=\$YABAI_WINDOW_ID &> /dev/null"
yabai -m signal --add event=window_focused action="sketchybar --trigger yabai_window_focused ID=\$YABAI_WINDOW_ID &> /dev/null"
yabai -m signal --add event=window_minimized action="sketchybar --trigger yabai_window_minimized ID=\$YABAI_WINDOW_ID &> /dev/null"
yabai -m signal --add event=window_deminimized action="sketchybar --trigger yabai_window_deminimized ID=\$YABAI_WINDOW_ID &> /dev/null"

```
bash
```
#!/usr/bin/env bash

yabai_get_windows_in_space() {
  local sid="$1"
  local result=()
  while IFS= read -r window_id; do
    result+=("$window_id")
  done < <(yabai -m query --spaces --space "$sid" | jq -r ".windows[]")

  # Return the result as a space-separated string
  echo "${result[@]}"
}

yabai_get_window_app_name() {
  local window_id="$1"
  yabai -m query --windows --window "$window_id" | jq -r '.app'
}

yabai_get_focused_window_id() {
  local window_id=$(yabai -m query --windows | jq -r '.[] | select(.["has-focus"] == true) | .id')
  if [[ -z "${window_id// }" ]]; then
    window_id=""
  fi
  echo "$window_id"
}
```
